### PR TITLE
Subscription authorizer for API

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*

*Description of changes:*
~~**Currently not working**~~
~~Ideally, `SubscriptionWithAwsIamTest` (identical to `GraphQLInstrumentationTest` except for the API being used) should pass, but it does not.~~

**WORKS NOW**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amplifyframework/api/graphql/GraphQLRequest.java [hunk 1]
```diff
@@ -69,6 +71,22 @@ public GraphQLRequest(
...
+     */
+    public String getDocument() {
+        return document;
+    }
+
+    /**
+     * Returns the immutable map of request variables.
+     * @return the map of variables
+     */
+    public Map<String, Object> getVariables() {
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/api/graphql/GraphQLRequest.java:86]: Ah heck. I was gonna say, "remember to add this to the `hashCode()`/`equals()`/`toString()`!" But it looks like we don't even have those in this class. Which, given how prominent it is in the API signatures, is bad news.

File: aws-api/src/test/java/com/amplifyframework/api/aws/SubscriptionAuthorizerTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    private ApiAuthProviders apiAuthProviders;
+
+    /**
+     * Construct fake auth providers to override default behaviors for
+     * subscription authorizer.
+     */
+    @Before
+    public void setup() {
+        authenticationSecret = RandomString.string();
+        apiAuthProviders = ApiAuthProviders.builder()
```
Comment (jamesonwilliams) [inline aws-api/src/test/java/com/amplifyframework/api/aws/SubscriptionAuthorizerTest.java:50]: I've never liked this `ApiAuthProviders` thing, it has poor encapsulation, and muddles together several different technologies without a consistent behavioral contract.

File: aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+    private URI getRequestEndpoint(boolean connectionFlag) throws ApiException {
+        try {
+            String baseUrl = configuration.getEndpoint();
+            String connectionUrl = connectionFlag ? baseUrl + "/connect" : baseUrl;
+            return new URI(connectionUrl);
+        } catch (URISyntaxException uriException) {
+            throw new ApiException(
+                    "Error constructing canonical URI for IAM request signature",
+                    uriException, AmplifyException.TODO_RECOVERY_SUGGESTION
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java:198]: Use the `REPORT_TO_AWS`-or-whatever one that @drochetti added

File: aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                credentialsProvider.getCredentials()
+        );
+
+        // Extract header from signed request and return
+        Map<String, String> signedHeaders = canonicalRequest.getHeaders();
+        JSONObject authorization = new JSONObject();
+        try {
+            for (Map.Entry<String, String> headerEntry : signedHeaders.entrySet()) {
+                authorization.put(headerEntry.getKey(), headerEntry.getValue());
+            }
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java:176]: Just, `authorization.putAll(signedHeaders()))`?

File: aws-api/src/androidTest/java/com/amplifyframework/api/aws/SubscriptionWithAwsIamTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
+/*
```
Comment (jamesonwilliams) [inline aws-api/src/androidTest/java/com/amplifyframework/api/aws/SubscriptionWithAwsIamTest.java:1]: How can you move this to `src/test`?

**Comment by drochetti** [comment]
> We should get rid of the `ApiAuthProviders`.
> 
> One way to do this is to have a different OkHttp `Interceptor` implementation for each type. Each implementation would know how to append headers into the request. This way, we'd have a standard behavioral contract for how auth gets added.
> 
> I guess there's that one weird detail around IAM and changing the endpoint. I'm not sure if you can rewrite the URL using an interceptor, but possibly.
> 

That's a good call and that's how we handle on iOS. The interceptor should be able to change anything in the request object given the request was not submitted yet.

One approach could be to pass a mutable builder around to the interceptors and once the interceptor chain is exhausted `build` is called to get an immutable request object

File: core/src/main/java/com/amplifyframework/api/graphql/GraphQLRequest.java (hunk 1 above)
Comment (raphkim) [reply core/src/main/java/com/amplifyframework/api/graphql/GraphQLRequest.java]: I removed `getVariables` and `getDocument` again after I realized that it isn't needed anymore. but I agree it still needs those methods you brought up.

File: aws-api/src/test/java/com/amplifyframework/api/aws/SubscriptionAuthorizerTest.java (hunk 2 above)
Comment (raphkim) [reply aws-api/src/test/java/com/amplifyframework/api/aws/SubscriptionAuthorizerTest.java:50]: Yea it was a poor implementation on my part back when I was working on it for mutate/query, but it's a legacy that should stay for until after GA...

File: aws-api/src/androidTest/java/com/amplifyframework/api/aws/SubscriptionWithAwsIamTest.java (hunk 5 above)
Comment (raphkim) [reply aws-api/src/androidTest/java/com/amplifyframework/api/aws/SubscriptionWithAwsIamTest.java]: This test was removed for now

File: aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java (hunk 4 above)
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java:176]: `JSONObject` doesn't have `putAll()` method, but I realized that I can just directly construct one with a map, so I'll change it.

File: aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java (hunk 3 above)
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionAuthorizer.java:198]: Didn't know we had that! I'll change it, and add appropriate recovery suggestion everywhere else.

**Pull request merged**